File: packages/agent/agent/lib/financial-actions.ts
export type InvoiceResolutionPlan = {
  differenceBeforeCents: number;
  adjustmentCents: number;
  differenceAfterExpectedCents: 0;
};

/**
 * Invariante contábil usada tanto na preparação quanto na execução.
 *
 * A diferença do checksum é `extraído - declarado`; o delta que fecha a conta
 * é necessariamente o inverso. Manter isso fora do modelo e fora da UI evita
 * que cada camada reinvente o sinal.
 */
export function invoiceResolutionPlan(
  differenceCents: number | null,
): InvoiceResolutionPlan | null {
  if (differenceCents === null || differenceCents === 0 || !Number.isInteger(differenceCents)) {
    return null;
  }
  return {
    differenceBeforeCents: differenceCents,
    adjustmentCents: -differenceCents,
    differenceAfterExpectedCents: 0,
  };
}

export function sameRevision(current: Date, expected: Date): boolean {
  return current.getTime() === expected.getTime();
}

File: services/agent/clara/ledger/financial_actions.py
"""Invariante contábil da resolução de divergência de fatura — porta de
`agent/lib/financial-actions.ts`.

Fica fora do modelo e fora de `prepare`/`apply` de propósito: manter o sinal
num lugar só evita que cada camada reinvente a conta.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class InvoiceResolutionPlan:
    difference_before_cents: int
    adjustment_cents: int
    difference_after_expected_cents: int = 0


def invoice_resolution_plan(difference_cents: int | None) -> InvoiceResolutionPlan | None:
    """A diferença do checksum é `extraído - declarado`; o ajuste que fecha
    a conta é necessariamente o inverso."""
    if difference_cents is None or difference_cents == 0:
        return None
    return InvoiceResolutionPlan(
        difference_before_cents=difference_cents,
        adjustment_cents=-difference_cents,
    )
